pkg/syslog-ng-ctl/stats_prometheus.go

```diff
@@ -19,132 +19,211 @@ import (
+				newLabel("result", stat.Type),
+			}
+			if stat.SourceInstance != "" {
+				labels = append(labels, newLabel("driver_instance", stat.SourceInstance))
+			}
+			if err := pushMetric(mfs, metric_ns+"_input_events_total", io_prometheus_client.MetricType_COUNTER, labels, float64(stat.Number)); err != nil {
+				errs = append(errs, err)
+			}
+		case strings.HasPrefix(stat.SourceName, "dst.") && slices.Contains([]string{"dropped", "queued", "written"}, stat.Type):
+			result := stat.Type
+			if result == "written" {
+				result = "delivered"
+			}
+			labels := []*io_prometheus_client.LabelPair{
+				newLabel("id", stat.SourceID),
+				newLabel("result", result),
 			}
-			if *mf.Type != io_prometheus_client.MetricType_UNTYPED {
-				continue
+			if stat.SourceInstance != "" {
+				labels = append(labels, newLabel("driver_instance", stat.SourceInstance))
 			}
-			if mf.Name == nil {
-				continue
+			if err := pushMetric(mfs, metric_ns+"_output_events_total", io_prometheus_client.MetricType_COUNTER, labels, float64(stat.Number)); err != nil {
+				errs = append(errs, err)
 			}
+		case stat.SourceName == "parser":
+			labels := []*io_prometheus_client.LabelPair{
+				newLabel("id", stat.SourceID),
+				newLabel("result", stat.Type),
+			}
+			if err := pushMetric(mfs, metric_ns+"_parsed_events_total", io_prometheus_client.MetricType_COUNTER, labels, float64(stat.Number)); err != nil {
+				errs = append(errs, err)
+			}
+		case stat.SourceName == "tag":
+			labels := []*io_prometheus_client.LabelPair{
+				newLabel("id", stat.SourceID),
+				newLabel("result", stat.Type),
+			}
+			if err := pushMetric(mfs, metric_ns+"_tagged_events_total", io_prometheus_client.MetricType_COUNTER, labels, float64(stat.Number)); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	err = errors.Join(errs...)
+	return mfs, err
+}
 
-			switch {
-			case strings.HasSuffix(*mf.Name, "_events_total"):
-				for _, m := range mf.Metric {
-					m.Counter = &io_prometheus_client.Counter{
-						Value: m.Untyped.Value,
-					}
-					m.Untyped = nil
+func metricCmp(A *io_prometheus_client.Metric, B *io_prometheus_client.Metric) int {
+	for i := range A.Label {
+		if i >= len(B.Label) {
+			return 1
+		}
+
+		lCmp := strings.Compare(*A.Label[i].Name, *B.Label[i].Name)
+		if lCmp != 0 {
+			return lCmp
+		}
+
+		lCmp = strings.Compare(*A.Label[i].Value, *B.Label[i].Value)
+		if lCmp != 0 {
+			return lCmp
+		}
+	}
+
+	return 0
+}
+
+func transformEventDelayMetric(delayMetric *io_prometheus_client.MetricFamily, delayMetricAge *io_prometheus_client.MetricFamily, now time.Time, lastMetricQueryTime time.Time, mfs map[string]*io_prometheus_client.MetricFamily) {
+	slices.SortFunc(delayMetric.Metric, metricCmp)
+	slices.SortFunc(delayMetricAge.Metric, metricCmp)
+
+	transformedMetric := []*io_prometheus_client.Metric{}
+	for i, m := range delayMetric.Metric {
+		delayMetric := m
+		delayMetricAge := delayMetricAge.Metric[i].GetGauge().GetValue()
```

Comment:
what about storing and looking up the age in a map (labels -> age) instead? it might be more efficient (not sure about that though) even if it's guaranteed that the two metrics will have their elements matching up if they are sorted the same way

Reply:
what I don't see is how much we can trust the source to ensure that if we sort the two metrics, their elements will match. if something is wrong there, we not only give incorrect results but the code will even panic for example if delayMetricAge.Metric is shorter than delayMetric.Metric.

2*N*sort should generally have a very similar complexity as N * (map set + lookup), if that counts at all in this case

Reply:
Sorry for the long delay.
It turned out that we can't deep-compare structs anymore, because a simple `fmt.Sprint(metric.Label)` call has side-effects on the Metric struct (it's a protobuf struct and it has caches as internal state).